x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/CoordinatorProxyAction.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.enrich.action;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.search.MultiSearchRequest;
+import org.elasticsearch.action.search.MultiSearchResponse;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.enrich.EnrichPlugin;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * An internal action to locally manage the load of the search requests the originate from the enrich processor.
+ * This is because the enrich processor executes in non blocking manner and a bulk request could easily overload
+ * the search tp.
+ */
+public class CoordinatorProxyAction extends ActionType<SearchResponse> {
+
+    public static final CoordinatorProxyAction INSTANCE = new CoordinatorProxyAction();
+    public static final String NAME = "indices:data/read/xpack/enrich/coordinate_lookups";
+
+    private CoordinatorProxyAction() {
+        super(NAME, SearchResponse::new);
+    }
+
+    public static class TransportAction extends HandledTransportAction<SearchRequest, SearchResponse> {
+
+        private final Coordinator coordinator;
+
+        @Inject
+        public TransportAction(TransportService transportService, ActionFilters actionFilters, Coordinator coordinator) {
+            super(NAME, transportService, actionFilters, (Writeable.Reader<SearchRequest>) SearchRequest::new);
+            this.coordinator = coordinator;
+        }
+
+        @Override
+        protected void doExecute(Task task, SearchRequest request, ActionListener<SearchResponse> listener) {
+            coordinator.schedule(request, listener);
+        }
+    }
+
+    public static class Coordinator {
+
+        final Client client;
+        final int maxLookupsPerRequest;
+        final int maxNumberOfConcurrentRequests;
+        final BlockingQueue<Item> queue = new LinkedBlockingQueue<>();
```

Comment:
Should we bound the size of the queue?

Reply:
Currently the capacity of this queue is `Integer.MAX_VALUE`. This queue will be shared by all enrich processor instances (on a node level). I do think it is okay for this queue to be able buffer a high number of search requests. What do you think will be a good number for this queue's capacity? Something like 100k?

Reply:
I think it is ok to buffer too, but I'd like us to have some guarantee that this can never cause out-of-memory issues? Maybe something in the order of `MAX_CONCURRENT_REQUESTS*MAX_LOOKUPS_PER_REQUEST` would make a good default? Something I'm not sure is whether we should wait or fail in case the queue is full.

Reply:
> Maybe something in the order of MAX_CONCURRENT_REQUESTS*MAX_LOOKUPS_PER_REQUEST would make a good default? 

That sounds good to me.

> Something I'm not sure is whether we should wait or fail in case the queue is full.

I'm leaning towards waiting. The `schedule(...)` method is indirectly called from the enrich processor on the same thread. If wait then there then we create natural back pressure. The client that sent the bulk requests waits until the response is returned. Ingestion happens on the write thread, so if the load is getting tool much on a cluster wide level than 429 will be returned.

Reply:
OK